src/ImageSharp.Formats.Tiff/TiffDecoderCore.cs

```diff
@@ -0,0 +1,452 @@
+        {
+            if (entry.Type != TiffType.SRational)
+                throw new ImageFormatException($"A value of type '{entry.Type}' cannot be converted to a SignedRational.");
+
+            byte[] bytes = ReadBytes(ref entry);
+            SignedRational[] result = new SignedRational[entry.Count];
+
+            for (int i = 0 ; i < result.Length ; i++)
+            {
+                int numerator = ToInt32(bytes, i * TiffConstants.SizeOfRational);
+                int denominator = ToInt32(bytes, i * TiffConstants.SizeOfRational + TiffConstants.SizeOfLong);
+                result[i] = new SignedRational(numerator, denominator);
+            }
+
+            return result;
+        }
+
+        public float ReadFloat(ref TiffIfdEntry entry)
+        {
+            if (entry.Count != 1)
+                throw new ImageFormatException($"Cannot read a single value from an array of multiple items.");
+
+            if (entry.Type != TiffType.Float)
+                throw new ImageFormatException($"A value of type '{entry.Type}' cannot be converted to a float.");
+
+            return ToSingle(entry.Value, 0);
+        }
+
+        public double ReadDouble(ref TiffIfdEntry entry)
+        {
+            if (entry.Count != 1)
+                throw new ImageFormatException($"Cannot read a single value from an array of multiple items.");
+
+            return ReadDoubleArray(ref entry)[0];
+        }
+
+        public float[] ReadFloatArray(ref TiffIfdEntry entry)
+        {
+            if (entry.Type != TiffType.Float)
+                throw new ImageFormatException($"A value of type '{entry.Type}' cannot be converted to a float.");
+
+            byte[] bytes = ReadBytes(ref entry);
+            float[] result = new float[entry.Count];
+
+            for (int i = 0 ; i < result.Length ; i++)
+                result[i] = ToSingle(bytes, i * TiffConstants.SizeOfFloat);
+
+            return result;
+        }
+
+        public double[] ReadDoubleArray(ref TiffIfdEntry entry)
+        {
+            if (entry.Type != TiffType.Double)
+                throw new ImageFormatException($"A value of type '{entry.Type}' cannot be converted to a double.");
+
+            byte[] bytes = ReadBytes(ref entry);
+            double[] result = new double[entry.Count];
+
+            for (int i = 0 ; i < result.Length ; i++)
+                result[i] = ToDouble(bytes, i * TiffConstants.SizeOfDouble);
+
+            return result;
+        }
+
+        private SByte ToSByte(byte[] bytes, int offset)
+        {
+            return (sbyte)bytes[offset];
+        }
+
+        private Int16 ToInt16(byte[] bytes, int offset)
+        {
+            if (IsLittleEndian)
+                return (short)(bytes[offset + 0] | (bytes[offset + 1] << 8));
+            else
+                return (short)((bytes[offset + 0] << 8) | bytes[offset + 1]);
+        }
+
+        private Int32 ToInt32(byte[] bytes, int offset)
+        {
+            if (IsLittleEndian)
```

Comment:
@JimBobSquarePants has introduced several classes to deal with endianness in streams. Shouldn't we use them 
in Tiff decoder too?
They definitely need some optimization by eliminating super-frequent virtual calls, but we could follow an "optimize once, win everywhere" approach by using them!